### PR TITLE
Donut acceleration project

### PR DESCRIPTION
**Pull request by tcal-x**
There is a README in the project directory.

### CODE REVIEW

File: proj/donut-accel/Makefile [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+# Software debugging
+# DEFINES += CFU_SOFTWARE_DEFINED
+
+# Uncommenting this moves donut code to "sram" region.  
+#   Probably not supported on all targets.
+# DEFINES += OPT_LINK_CODE_IN_SRAM
+
+
+export EXTRA_LITEX_ARGS
+EXTRA_LITEX_ARGS += --cpu-variant breaker+cfu
```
Comment (ShvetankPrakash) [inline proj/donut-accel/Makefile:43]: Why again do we need this variant specifically?

Comment (ShvetankPrakash) [reply proj/donut-accel/Makefile:43]: Ah, it's cause we want the slower iterative multiplier on icebreaker, right? @tcal-x

Comment (tcal-x) [reply proj/donut-accel/Makefile:43]: > Ah, it's cause we want the slower iterative multiplier on icebreaker, right? @tcal-x

Yes, only used in this project and with Icebreaker.

File: proj/donut-accel/src/donut.c [hunk 2]
```diff
@@ -0,0 +1,87 @@
+// The donut code with fixed-point arithmetic; no sines, cosines, square roots, or anything.
+// a1k0n 2020
+// Code from: https://gist.github.com/a1k0n/80f48aa8911fffd805316b8ba8f48e83
+// For more info:
+// - https://www.a1k0n.net/2011/07/20/donut-math.html
+// - https://www.youtube.com/watch?v=DEqXNfs_HhY
+
```
Comment (ShvetankPrakash) [inline proj/donut-accel/src/donut.c:7]: Since this is the accelerated version maybe it would be nice if we like make the accelerated instructions standout more with caps (or a comment or something)? Just thinking at like HPCA if we walk thru this code (or if anyone is walking thru this as a tutorial example) it would be nice for people to easily see what instructions are the accelerated ones by glance.

File: proj/donut-accel/src/software_cfu.cc [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+#include <stdint.h>
+#include "software_cfu.h"
+
+//
+// In this function, place C code to emulate your CFU. You can switch between
+// hardware and emulated CFU by setting the CFU_SOFTWARE_DEFINED DEFINE in
+// the Makefile.
+uint32_t software_cfu(int funct3, int funct7, uint32_t rs1, uint32_t rs2)
+{
+  //return (funct3 & 1) ? 0 : ((rs1 * rs2) >> 10);
```
Comment (ShvetankPrakash) [inline proj/donut-accel/src/software_cfu.cc:26]: Nit: can we remove this comment?

File: third_party/litex-donut/donut.c [hunk 4]
```diff
@@ -21,18 +22,25 @@
...
   puts("\nPress any key to exit.\n");
   for (;;) {
+    unsigned start_cycle = perf_get_mcycle();
     memset(b, 32, 1760);  // text buffer
     memset(z, 127, 1760);   // z buffer
     int sj=0, cj=1024;
     for (int j = 0; j < 90; j++) {
       int si = 0, ci = 1024;  // sine and cosine of angle i
       for (int i = 0; i < 324; i++) {
-        int R1 = 1, R2 = 2048, K2 = 5120*1024;
```
Comment (ShvetankPrakash) [inline third_party/litex-donut/donut.c:35]: Why did you move this line?

File: proj/donut-accel/src/software_cfu.cc (hunk 3 above)
Comment (tcal-x) [reply proj/donut-accel/src/software_cfu.cc:26]: Done

File: proj/donut-accel/src/donut.c (hunk 2 above)
Comment (tcal-x) [reply proj/donut-accel/src/donut.c:7]: I've capitalized the custom macros.

File: third_party/litex-donut/donut.c (hunk 4 above)
Comment (tcal-x) [reply third_party/litex-donut/donut.c]: I moved it back to avoid confusion (also in the accelerated version).

**Comment by tcal-x** [comment]
@ShvetankPrakash PTAL

**Comment by ShvetankPrakash** [comment]
I verified on the board and LGTM! The difference is a bit hard to see with the naked eye, no? I can easily see from the cycle count being printed though between the accelerated and non accelerated versions which is really nice :)

**Review by ShvetankPrakash — APPROVED** [review]

**Comment by ShvetankPrakash** [comment]
@tcal-x you fine with merging this now? Will want this for HPCA :)

**Comment by tcal-x** [comment]
I'd like to review the speedup or lack thereof.   I thought I saw more speedup than you did.
